### PR TITLE
Migrate to operatorkit with legacy resource and stub cf resource

### PR DESCRIPTION
**Pull request by rossf7**
Towards giantswarm/giantswarm#2023

This PR does the migration to operatorkit with a legacy resource and the stub cloud formation resource. Once it's approved I'll convert the operatorkit branch into a protected branch rather than merging to master.

Once we have both the legacy and first version of the cloud formation resource fully working we can merge the branch into master. For the rest of the cloud formation migration we can then merge into master.

With the legacy resource there is 1 problem with scaling worker nodes. Now there isn't an updateFunc so we need to update the ASG size when re-processing clusters. I'd like to fix that as a separate change.

### CODE REVIEW

File: service/messagecontext/message_context.go [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+// this key directly.
+var messageKey key = "message"
+
+// Message is a communication structure used to transport information from one
+// resource to another. Messages move between resource during reconciliation
+// within the dispatched context.
+type Message struct {
+	// ConfigMapNames is a list of config map names filled by the config map
+	// resource and read by the deployment resource.
+	ConfigMapNames []string
```
Comment (xh3b4sd) [inline service/messagecontext/message_context.go]: What is the message context used for? I do not see it being used and the config map names here should be irrelevant as well.

File: service/operator/operator.go [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+
+		c.Description = awstpr.Description
+		c.Name = awstpr.Name
+		c.Version = awstpr.VersionV1
+
+		newTPR, err = tpr.New(c)
+		if err != nil {
+			return nil, microerror.Maskf(err, fmt.Sprintf("creating TPR util for %s", awstpr.Name))
+		}
+	}
```
Comment (xh3b4sd) [inline service/operator/operator.go]: Current standard should be something like this: https://github.com/giantswarm/kvm-operator/blob/7bf46cd0c5130d104516c854c420d78b99117eef/service/operator/operator.go. E.g. we use the operatorkit informer and also inject the TPR now.

File: service/resource/cloudformation/resource.go [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+			"resource", Name,
+		),
+	}
+
+	return newService, nil
+}
+
+func (r *Resource) GetCurrentState(ctx context.Context, obj interface{}) (interface{}, error) {
+	return nil, nil
+}
```
Comment (xh3b4sd) [inline service/resource/cloudformation/resource.go]: It would be cool to separate the steps into multiple files. We can do this in a separate PR though.

File: service/resource/legacy/resource.go [hunk 4]
```diff
@@ -45,7 +49,7 @@ const (
 	gracePeriodSeconds = 10
 )
 
-// Config represents the configuration used to create a version service.
+// Config represents the configuration used to create a legacy service.
```
Comment (xh3b4sd) [inline service/resource/legacy/resource.go]: FYI in some places we started to omit these godocs.

File: service/resource/legacy/resource.go [hunk 5]
```diff
@@ -173,57 +177,34 @@ type Event struct {
...
-
-func (s *Service) addFunc(obj interface{}) {
+// NewUpdatePatch is called upon observed custom object change. It creates the
+// AWS resources for the cluster.
+func (s *Resource) NewUpdatePatch(ctx context.Context, obj, currentState, desiredState interface{}) (*framework.Patch, error) {
 	customObject, ok := obj.(*awstpr.CustomObject)
 	if !ok {
-		s.logger.Log("error", "could not convert to awstpr.CustomObject")
-		return
+		return &framework.Patch{}, microerror.Maskf(invalidConfigError, "could not convert to awstpr.CustomObject")
```
Comment (xh3b4sd) [inline service/resource/legacy/resource.go]: Zero value of a pointer is `nil`. Makes the return value easier.

File: service/service.go [hunk 6]
```diff
@@ -146,19 +166,98 @@ func New(config Config) (*Service, error) {
...
+			return nil, microerror.Mask(err)
+		}
+	}
+
+	var cloudformationResource framework.Resource
+	{
+		cloudformationConfig := cloudformationresource.DefaultConfig()
+
+		cloudformationConfig.CertWatcher = certWatcher
+		//cloudformationConfig.CloudConfig = ccService
```
Comment (xh3b4sd) [inline service/service.go]: What happens with that?

File: service/resource/cloudformation/resource.go (hunk 3 above)
Comment (fgimenez) [reply service/resource/cloudformation/resource.go]: Yep, this is coming here https://github.com/giantswarm/aws-operator/pull/484

File: service/resource/legacy/resource.go [hunk 7]
```diff
@@ -252,56 +231,62 @@ func (s *Service) deleteFunc(obj interface{}) {
...
+	return &framework.Patch{}, nil
 }
 
-// TODO we need to support this in operatorkit.
-func (s *Service) updateFunc(oldObj, newObj interface{}) {
-	oldCluster := *oldObj.(*awstpr.CustomObject)
-	cluster := *newObj.(*awstpr.CustomObject)
+func (s *Resource) ApplyUpdateChange(ctx context.Context, obj, updateChange interface{}) error {
+	// TODO The legacy resource needs to support scaling workers via NewUpdatePatch.
+	/*
```
Comment (fgimenez) [inline service/resource/legacy/resource.go]: Do we need to keep this commented code?

File: service/messagecontext/message_context.go (hunk 1 above)
Comment (rossf7) [reply service/messagecontext/message_context.go]: This was in the migration branch and I just copied it across. My bad, it's now removed.

File: service/resource/cloudformation/resource.go (hunk 3 above)
Comment (rossf7) [reply service/resource/cloudformation/resource.go]: Yes, I like the separating into multiple files. Let's do in a later PR.

File: service/resource/legacy/resource.go (hunk 5 above)
Comment (rossf7) [reply service/resource/legacy/resource.go]: Good point. I'm returning nil in all except the final return.

File: service/service.go (hunk 6 above)
Comment (rossf7) [reply service/service.go]: This was also from the migration branch. The cloudconfig service is now passed in for the stub.

**Comment by rossf7** [comment]
@xh3b4sd PTAL. The operator service is now based on the latest example you gave me. I'll also remove your migration branch when I protect this new one.

**Comment by rossf7** [comment]
@fgimenez @xh3b4sd FYI. The operatorkit branch is now protected and can be used for the migration PRs. I've also deleted the old branch and closed the PRs.
